### PR TITLE
Rework output

### PR DESCRIPTION
**Pull request by wence-**
Redo viz output to simplify things and make supporting interpolated output easier.  A side effect of this is that we no longer require evtk in the tree (since we're just pasting a little bit of xml).

We deprecate the `<<` output syntax (with loud warnings) in favour of:

```python
output.write(*args, **kwargs)
```

Particularly, file output now supports multiple functions in the same file.  For example, to save both the velocity and pressure in a time loop (remembering the timestep).

```python
while t < ... :

    ...
    output.write(velocity, pressure, time=t)
```

File output can decide whether or not the fields are `project`ed or `interpolate`d to the output solution points.  The default is interpolation.  This can be selected when building the object:

```python
output = File("foo.pvd", project_output=True) # output will be projected if necessary
output = File("foo.pvd") # output will be interpolated if necessary
```

Interpolation/projection only happens for those functions which need it.  The determination of this has also been simplified by looking at the fiat element.

Additionally, refactor the interpolation functionality attached to Function objects into a separate file, and add an `Interpolator` (analogous to a `Projector`) that returns an object that can reusably interpolate an expression (useful in a timestepping loop, and also used by the File output).

I've updated the demos to conform to the new API.

### CODE REVIEW

File: demos/burgers/burgers.py.rst [hunk 1]
```diff
@@ -90,7 +90,7 @@ To output the initial conditions, we project them into P1 and pipe
 them into the output file::
 
   outfile = File("burgers.pvd")
-  outfile << project(u, V_out, name="Velocity")
+  outfile.write(project(u, V_out, name="Velocity"))
```
Comment (miklos1) [inline demos/burgers/burgers.py.rst]: I see the original projects as well, so just asking: can't we use `project_output` here?

File: demos/camassa-holm/camassaholm.py.rst [hunk 2]
```diff
@@ -206,7 +206,7 @@ Finally, we check if it is time to dump the data. ::
      dumpn += 1
      if dumpn == ndump:
         dumpn -= ndump
-        ufile << u1
+        ufile.write(u1, time=t)
```
Comment (miklos1) [inline demos/camassa-holm/camassaholm.py.rst]: So we can now save more information (the time) than previously?

File: demos/linear-wave-equation/linear_wave_equation.py.rst [hunk 3]
```diff
@@ -67,7 +67,7 @@ output file::
 Output the initial conditions::
 
   outfile = File("out.pvd")
-  outfile << phi
+  outfile.write(phi)
```
Comment (miklos1) [inline demos/linear-wave-equation/linear_wave_equation.py.rst]: We use no `time=0` for the initial condition, but use `time=t` for all the other time steps. What will we see in Paraview with respect to time?

File: demos/camassa-holm/camassaholm.py.rst (hunk 2 above)
Comment (wence-) [inline demos/camassa-holm/camassaholm.py.rst]: There was previously some magic format you could pass the data in to `<<` that allowed you to specify the timestep.  But I don't think anyone knew what it was, or used it.

File: demos/burgers/burgers.py.rst (hunk 1 above)
Comment (wence-) [inline demos/burgers/burgers.py.rst]: Yes.

File: demos/linear-wave-equation/linear_wave_equation.py.rst (hunk 3 above)
Comment (wence-) [inline demos/linear-wave-equation/linear_wave_equation.py.rst]: `time` defaults to an integer counting from zero.  I should document that.

File: firedrake/output.py [hunk 4]
```diff
@@ -1,562 +1,481 @@
...
+VTK_TRIANGLE = 5
+VTK_QUADRILATERAL = 9
+VTK_TETRAHEDRON = 10
+VTK_HEXAHEDRON = 12
+VTK_WEDGE = 13
 
-    .. code-block:: python
+cells = {
+    ufl.Cell("interval"): VTK_INTERVAL,
+    ufl.Cell("triangle"): VTK_TRIANGLE,
```
Comment (miklos1) [inline firedrake/output.py]: So the tdim != gdim cases seem to have disappeared.

Comment (wence-) [inline firedrake/output.py]: I get the cell from the mesh topology.

File: firedrake/output.py [hunk 5]
```diff
@@ -1,562 +1,481 @@
...
+            "|": "LittleEndian",
+            "<": "LittleEndian",
+            ">": "BigEndian"}[dtype.byteorder]
+
+
+def write_array(f, ofunction):
+    array = ofunction.array
+    numpy.uint32(array.nbytes).tofile(f)
+    if get_byte_order(array.dtype) == "BigEndian":
+        array = array.byteswap()
```
Comment (miklos1) [inline firedrake/output.py]: So the disk format is always little endian?

Comment (wence-) [inline firedrake/output.py]: Yes.  I could probably just remove this entirely and hard-code little endian, given that I don't know of anyone running firedrake on ppc, but this is reasonably safe I think.

File: firedrake/output.py [hunk 6]
```diff
@@ -1,562 +1,481 @@
...
+                              mode="constant")
+    else:
+        raise ValueError("Can't write data with shape %s" % shape)
+    return array
+
+
+class File(object):
+    header = ('<?xml version="1.0" ?>\n'
+              '<VTKFile type="Collection" version="0.1" '
+              'byte_order="LittleEndian">\n'
```
Comment (miklos1) [inline firedrake/output.py]: Ah, I see what's going on.

**Comment by miklos1** [comment]
Maybe we could have some documentation on this?

**Comment by miklos1** [comment]
I think this is overall fine.

**Comment by wence-** [comment]
Added docs and fixed up a few minor issues.

**Comment by miklos1** [comment]
OK.

**Comment by wence-** [comment]
fixedified.

File: firedrake/output.py [hunk 7]
```diff
@@ -1,562 +1,502 @@
...
+                    raise ValueError("Can't write tensors with more than 9 components")
+                V = TensorFunctionSpace(function.ufl_domain(), family, 1,
+                                        shape=shape)
             else:
-                raise TypeError(
-                    "You cannot start setting the time by giving a tuple.")
+                raise ValueError("Unsupported shape %s" % (shape, ))
+            output = Function(V)
+            self._output_functions[function] = output
+
```
Comment (tkarna) [inline firedrake/output.py]: This allocates the output function, which wastes memory if there are many functions (or File objects) that use the same output space V.

File: firedrake/output.py [hunk 8]
```diff
@@ -1,562 +1,502 @@
...
+                # Write new dataset name
+                f.write('<DataSet timestep="%s" '
+                        'file="%s" />\n' % (time, vtu))
+                # And add footer again, so that the file is valid
+                f.write(self._footer)
+
+    def __lshift__(self, arg):
+        from pyop2.logger import warning, RED
+        warning(RED % "The << syntax is deprecated, use File.write")
+        self.write(arg)
```
Comment (tkarna) [inline firedrake/output.py]: __lshift__ no longer supports `File << (function, time)` construct.

**Comment by tkarna** [comment]
Would it make sense to have `Projector.apply` and `Interpolator.apply` instead of `Projector.project` and `Interpolator.interpolate` so that you wouldn't need to check the instance when using these operators?

**Comment by dham** [comment]
Or even implement `__call__` on `Projector` and `Interpolator` to do this.

**Pull request merged**